### PR TITLE
fix(caching): add config for number of concurrent scheduled agents

### PR DESCRIPTION
**Pull request by emjburns**
(Please view with ["ignore whitespace"](https://blog.github.com/2011-10-21-github-secrets/#whitespace)).

We just saw clouddriver try and acquire all the locks, then die. This makes the max number of caching agents to schedule at once and the interval at which we schedule configurable. It defaults to the current behavior.

Set with:
```
redis:
  agent:
    maxConcurrentAgents: 100
    agentLockAcquisitionIntervalSeconds: 5
```

### CODE REVIEW

File: cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredAgentScheduler.java [hunk 1]
```diff
@@ -43,240 +39,260 @@
...
+  private final RedisClientDelegate redisClientDelegate;
+  private final NodeIdentity nodeIdentity;
+  private final AgentIntervalProvider intervalProvider;
+  private final ExecutorService agentExecutionPool;
+  private final Pattern enabledAgentPattern;
+
+  private final Map<String, AgentExecutionAction> agents = new ConcurrentHashMap<>();
+  private final Map<String, NextAttempt> activeAgents = new ConcurrentHashMap<>();
+  private final NodeStatusProvider nodeStatusProvider;
+  private Integer maxConcurrentAgents;
```
Comment (cfieber) [inline cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredAgentScheduler.java]: `final`

File: cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredAgentScheduler.java [hunk 2]
```diff
@@ -43,240 +39,260 @@
...
+    Collections.shuffle(agentsEntrySet);
+    for (Map.Entry<String, AgentExecutionAction> agent : agentsEntrySet) {
+      if (!skip.contains(agent.getKey())) {
+        final String agentType = agent.getKey();
+        AgentIntervalProvider.Interval interval = intervalProvider.getInterval(agent.getValue().getAgent());
+        if (acquireRunKey(agentType, interval.getTimeout())) {
+          acquired.put(agentType, new NextAttempt(System.currentTimeMillis(), interval.getInterval(), interval.getErrorInterval()));
         }
+      }
+      if (acquired.keySet().size() >= availableAgents) {
```
Comment (cfieber) [inline cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredAgentScheduler.java]: `acquired.size()` should work

File: cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredAgentScheduler.java [hunk 3]
```diff
@@ -43,240 +39,260 @@
...
+                                 String enabledAgentPattern,
+                                 Integer maxConcurrentAgents,
+                                 Integer agentLockAcquisitionIntervalSeconds) {
+    this.redisClientDelegate = redisClientDelegate;
+    this.nodeIdentity = nodeIdentity;
+    this.intervalProvider = intervalProvider;
+    this.nodeStatusProvider = nodeStatusProvider;
+    this.agentExecutionPool = agentExecutionPool;
+    this.enabledAgentPattern = Pattern.compile(enabledAgentPattern);
+    this.maxConcurrentAgents = maxConcurrentAgents == null ? Integer.MAX_VALUE : maxConcurrentAgents;
```
Comment (robzienert) [inline cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredAgentScheduler.java]: Do we really want to have `Integer.MAX_VALUE` as the default? If we're going to have a default, I don't think it should be a value that we've demonstrated as being pathological.

Comment (emjburns) [reply cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredAgentScheduler.java]: The goal is to keep the behavior the same. `Integer.MAX_VALUE` is definitely too high ... but it does keep the behavior the same. Thoughts on changing the behavior to something more sensible? 100?

Comment (asher) [reply cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredAgentScheduler.java]: ➕

Comment (robzienert) [reply cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredAgentScheduler.java]: That works for me.

**Pull request merged**